Issue #8099: Add binary:join/2 to stdlib

onno-vos-dev opened the pull request:
See linked issue for details.

seriyps commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,28 @@ unhex(X) ->
+-spec join([binary()], binary()) -> binary().
+join([H], _Separator) -> H;
+join([H | T], Separator) ->
+    join(T, Separator, H).
```
as an option
```suggestion
    lists:foldl(fun(Element, Acc) -> <<Acc/binary, Separator/binary, Element/binary>> end, H, T).
```

mikpe replied:
Would that not lose some performance though? Performance being the main reason for this implementation.

mikpe commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,28 @@ unhex(X) ->
+""".
+-doc(#{since => <<"OTP 27.0">>}).
+-spec join([binary()], binary()) -> binary().
+join([H], _Separator) -> H;
```
`lists:join/2` accepts an empty list of things, I think `binary:join/2` should do the same.

onno-vos-dev commented on lib/stdlib/test/binary_module_SUITE.erl:
```diff
@@ -260,6 +260,9 @@ badargs(Config) when is_list(Config) ->
     badarg = ?MASK_ERROR(binary:encode_hex(foo)),
+
+    badarg = ?MASK_ERROR(binary:join([<<"a">>], ", ")),
+    badarg = ?MASK_ERROR(binary:join([], <<",">>)),
```
I think suspect might be a bit too gentle wording here 😄 You're right!

michalmuskala commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,29 @@ unhex(X) ->
+join([], _Separator) -> <<>>;
+join([H], _Separator) -> H;
+join([H | T], Separator) ->
+    join(T, Separator, H).
```
This can be improved to convince the compiler to use the new "private append" optimisation by making sure we start appending from an empty binary.
In my very simple & naive benchmarks this makes this ~20% faster on the data I used - it likely needs to be tested more thoroughly 
```suggestion
    Acc = <<>>,
    join(T, Separator, <<Acc/binary, H/binary>>).
```

onno-vos-dev replied:
Getting some mixed results on my benchmark (it's also a very naive benchmark 😄) but it does indeed consistently perform better with anywhere between 20-50% improvement 😮 👍

seriyps replied:
Might be nice to add a comment in the code explaining why this is needed.

garazdawi commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,30 @@ unhex(X) ->
             }).
 
+-doc """
+Joins a list of binaries together by a specified Separator.
+Equivalent to `iolist_to_binary(lists:join(Separator, Binaries))` with binary:join/2 being the faster of the two.
```
```suggestion
Joins a list of binaries together by a specified `Separator`.

Equivalent to `iolist_to_binary(lists:join(Separator, Binaries))` with `binary:join/2` being the faster of the two.
```

garazdawi replied:
Just to clarify that the extra newline was intentional. The first paragraph is used as a summary for the function and (atleast IMO) the first sentence in this case is enough of a summary.

onno-vos-dev replied:
Fixed once and for all now 😄

onno-vos-dev commented:
Squashed to get rid of some of the commit noise

Maria-12648430 commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,31 @@ unhex(X) ->
+-doc(#{since => <<"OTP 27.0">>}).
+-spec join([binary()], binary()) -> binary().
+join([], _Separator) -> <<>>;
+join([H], _Separator) -> H;
```
```suggestion
join([H], _Separator) when is_binary(H) -> H;
```

onno-vos-dev replied:
I noticed the same bug 😄 Had a test case that failed on me 😄

Maria-12648430 replied:
Now that I think about it, you should also check the `Separator` in the empty and single-element list clauses :sweat_smile:

onno-vos-dev replied:
I thought about it too but technically speaking it doesn't change the outcome. So while a `badarg` should be raised from that perspective, the outcome being `<<>> | H` wouldn't make use of the `Separator` so technically it's an unnecessary guard with little gain IMHO. 

Happy to include it though if more agree 👍

michalmuskala replied:
Actually it might make a small difference in performance to validate the type of separator before the main loop - the compiler will likely omit those typechecks later on each loop iteration

Maria-12648430 commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,31 @@ unhex(X) ->
+-spec join([binary()], binary()) -> binary().
+join([], _Separator) -> <<>>;
+join([H], _Separator) -> H;
+join([H | T], Separator) ->
+    Acc = <<>>,
+    join(T, Separator, <<Acc/binary, H/binary>>).
```
```suggestion
join([H | T], Separator) ->
    join(T, Separator, H).
```

onno-vos-dev replied:
@Maria-12648430 This change was specifically done based on earlier discussions with @michalmuskala 👍 It's an optimisation that I was at least not aware of but seems to make a nice difference. https://github.com/erlang/otp/pull/8100#pullrequestreview-1869763175

Maria-12648430 replied:
Ok, news to me :sweat_smile: Never mind then :+1:

paulo-ferraz-oliveira commented:
Fwiw, there was an attempt at this a [few years back](https://github.com/erlang/otp/pull/427), that ended up not making it due to a [core team decision](https://github.com/erlang/otp/pull/427#issuecomment-55573359). It's possible, though, that stuff's changed since then.

josevalim commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,34 @@ unhex(X) ->
+join([], _Separator) -> <<>>;
+join([H], Separator) when is_binary(H) andalso is_binary(Separator) -> H;
+join([H | T], Separator) when is_binary(Separator) ->
+    %% Starting with an empty binary convinces the compiler to use the new "private append" optimisation
+    Acc = <<>>,
+    join(T, Separator, <<Acc/binary, H/binary>>);
```
@bjorng @jhogberg @michalmuskala 

I understand why starting with the empty string makes the compiler use the new "private append" optimisation. However I wonder if this could be generalized?

While `H` cannot be privately appended because it comes from an external function, the next iteration of `join/3` can be privately appended, because no one uses the intermediate binary. Therefore, for those "closed loops", would it make sense to have a bit in the binary that tells when to private append or not? Generally speaking, everything after the first iteration would be privately appended. This would allow private append to happen in more situations, although I am not aware of the costs of reserving one extra bit for binaries.

PS: I may be completely off mark here.

bjorng replied:
> I understand why starting with the empty string makes the compiler use the new "private append" optimisation. However I wonder if this could be generalized?
> 
> While `H` cannot be privately appended because it comes from an external function, the next iteration of `join/3` can be privately appended, because no one uses the intermediate binary.

The private append operation is faster because it does fewer tests and less working than general append operation. Therefore, the private append **must** only be used with a binary that has been specially prepared.

However, it should be possible for the compiler to generalize the optimization. If the call looks like:

```
join(T, Separator, H)
```

the compiler could rewrite it to:

```
    Acc = <<>>,
    join(T, Separator, <<Acc/binary, H/binary>>);
```

Not sure that this code pattern is common enough to make the optimization worthwhile to implement, though. It would not be needed if the clause is rewritten to:

```
join([_ | _]=List, Separator) when is_binary(Separator) ->
    join(List, Separator, <<>>);
```

bjorng commented on lib/stdlib/src/binary.erl:
```diff
@@ -934,6 +935,34 @@ unhex(X) ->
+-doc(#{since => <<"OTP 27.0">>}).
+-spec join([binary()], binary()) -> binary().
+join([], _Separator) -> <<>>;
+join([H], Separator) when is_binary(H) andalso is_binary(Separator) -> H;
```
```suggestion
join([H], Separator) when is_binary(H), is_binary(Separator) -> H;
```

bjorng commented:
After the first release candidate, we generally focus on bug fixes and polishing of features already included or planned for the release. To ensure that Erlang/OTP 27 will be as good as it possibly can be, we need to minimize the time we spend on things not to be included in the release. Therefore, we will not review this pull request until after the OTP 27 has been released. If we have not came back to it before September, feel free to remind us.

bjorng commented:
Now we are back working on Erlang/OTP 28. OTP Technical Board will have to it approve this addition to the `binary` module. It is likely that it will be approved within a week or two.

Meanwhile, I've pushed a commit with suggested clean ups and proper error handling. Please review. If you approve, please rebase on the latest `master` and squash into one commit.

onno-vos-dev commented:
> Now we are back working on Erlang/OTP 28. OTP Technical Board will have to it approve this addition to the `binary` module. It is likely that it will be approved within a week or two.
> 
> Meanwhile, I've pushed a commit with suggested clean ups and proper error handling. Please review. If you approve, please rebase on the latest `master` and squash into one commit.

Thank you! Squashed and rebased now 👍

bjorng commented:
Thanks! Added to our daily builds.

bjorng commented:
Thanks for the pull request.
